text/0000-cfg-oneof.md

```diff
@@ -0,0 +1,97 @@
+
+The desired outcome is the ability to specify `#[cfg(oneof(feature = "a", feature = "b"))] do_something()` and `#[cfg(not(oneof(feature = "a", feature = "b")))] compile_error!(...)` to specify exclusive features without manually defining all possible valid/invalid combinations for the exclusive subset of features, and simplify other configurations that benefit from the `oneof` predicate.
+This allows authors of crates with exclusive feature sets specify this in a maintainable manner, and to communicate this to consumers without requiring users to infer the feature issue from compiler errors.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Additions to existing [guide](https://doc.rust-lang.org/reference/conditional-compilation.html)
+
+``` 
+ConfigurationOneof
+   oneof ( ConfigurationPredicateList )
+   
+...
+
+oneof() with a comma separated list of configuration predicates. It is true if at least one and only one predicate is true. In all other situations it is false.
+```
+
+## An Example
+
+Using an existing implementation from [rust-rand-facade](https://github.com/ryankurte/rust-rand-facade/blob/master/src/lib.rs) expressing all possible combinations for a set of three exclusive features (`std`, `cortex_m`, and `os_rng`). This is a reasonably small example, it is worth noting that it is common for embedded crates to have [tens of exclusive features](https://github.com/stm32-rs/stm32f4xx-hal/blob/master/Cargo.toml#L67).
+
+Using existing predicates:
+
+```rust
+#[cfg(all(feature = "std", feature = "cortex_m"))]
+compile_error!("Only one of 'std', 'os_rng', or 'cortex_m' features may be enabled");
+
+#[cfg(all(feature = "std", feature = "os_rng"))]
+compile_error!("Only one of 'std', 'os_rng', or 'cortex_m' features may be enabled");
+
+#[cfg(all(feature = "cortex_m", feature = "os_rng"))]
+compile_error!("Only one of 'std', 'os_rng', or 'cortex_m' features may be enabled");
+
+#[cfg(not(any(feature = "std", feature = "cortex_m", feature = "os_rng")))]
+compile_error!("One of 'os_rng', 'std', 'cortex_m' features must be enabled");
+```
+
+Using the `oneof` predicate:
+```rust
+#[cfg(not(oneof(feature = "std", feature = "cortex_m", feature = "os_rng")))]
+compile_error!("One of 'os_rng', 'std', 'cortex_m' features must be enabled");
+```
+
+The latter is significantly more simple, and does not introduce additional complexity when adding further exclusive feature(s).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+- I do not believe this has significant interaction with other features, or requires any significant changes to the tooling or language
+- Implementation should be straightforward, requiring the addition of a config predicate to `rustc` based on existing configuration predicates
+- I do not believe there are significant corner cases to be considered
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+N/A
+
+# Rationale and alternatives
```

Comment:
Possible alternative would be providing a `oneof!` macro which implements the behavour rather than a proper item, as this could be implemented with existing macros.

Reply:
@clarfon True, but I think that would be much less efficient than a native implementation.

Reply:
That seems like a pretty brave claim, honestly. Efficiency is not clearly defined here and I can't imagine a single use case where the difference would be noticeable to the end user.

Reply:
@clarfon As far as I can tell, the expansion would be quadratic in the number of conditions.

Reply:
And quadratic isn't really that bad. Computers are fast and the number of terms is low. If you can find a case where the compile time difference is on a noticeable scale then I'd say it's reasonable, but I can't imagine that ever being the case.